compiler/src/dotty/tools/dotc/transform/sjs/PrepJSInterop.scala

```diff
@@ -0,0 +1,1221 @@
+        }
+      }
+
+      if (sym.hasAnnotation(defn.NativeAnnot)) {
+        // Native methods are not allowed
+        report.error("Methods in a js.Any may not be @native", tree)
+      }
+
+      /* In native JS types, there should not be any private member, except
+       * private[this] constructors.
+       */
+      if ((enclosingOwner is OwnerKind.JSNative) && isPrivateMaybeWithin(sym)) {
+        if (sym.isClassConstructor) {
+          if (!sym.isAllOf(PrivateLocal)) {
+            report.error(
+                "Native JS classes may not have private constructors. " +
+                "Use `private[this]` to declare an internal constructor.",
+                sym)
+          }
+        } else if (!sym.is(ParamAccessor)) {
+          report.error(
+              "Native JS classes may not have private members. " +
+              "Use a public member in a private facade instead.",
+              tree)
+        }
+      }
+
+      if (enclosingOwner is OwnerKind.JSNonNative) {
+        // Private methods cannot be overloaded
+        if (sym.is(Method) && isPrivateMaybeWithin(sym)) {
+          val alts = sym.owner.info.memberBasedOnFlags(sym.name, required = Method)
+          if (alts.isOverloaded) {
+            report.error(
+                "Private methods in non-native JS classes cannot be overloaded. Use different names instead.",
+                tree)
+          }
+        }
+
+        // private[Scope] methods must be final
+        if (!sym.isOneOf(Final | Protected) && sym.privateWithin.exists && !sym.isClassConstructor)
+          report.error("Qualified private members in non-native JS classes must be final", tree)
+
+        // Traits must be pure interfaces, except for js.undefined members
+        if (sym.owner.is(Trait) && sym.isTerm && !sym.isConstructor) {
+          if (sym.is(Method) && isPrivateMaybeWithin(sym)) {
+            report.error("A non-native JS trait cannot contain private members", tree)
+          } else if (sym.is(Lazy)) {
+            report.error("A non-native JS trait cannot contain lazy vals", tree)
+          } else if (!sym.is(Deferred)) {
+            /* Tell the back-end not to emit this thing. In fact, this only
+             * matters for mixed-in members created from this member.
+             */
+            sym.addAnnotation(jsdefn.JSOptionalAnnot)
+
+            if (!sym.isSetter) {
+              // Check that methods do not have parens
+              if (sym.is(Method, butNot = Accessor) && sym.info.stripPoly.isInstanceOf[MethodType])
+                report.error("In non-native JS traits, defs with parentheses must be abstract.", tree.rhs)
+
+              // Check that the rhs is `js.undefined`
+              tree.rhs match {
+                case sel: Select if sel.symbol == jsdefn.JSPackage_undefined =>
+                  // ok
+                case Apply(Apply(TypeApply(fromTypeConstructorFun, _), (sel: Select) :: Nil), _) if sel.symbol == jsdefn.JSPackage_undefined =>
```

Comment:
Should check that the symbol of `fromTypeConstructorFun` is the symbol of `js.|.fromTypeConstructor`